src/TriggerBinding/SqlTriggerListener.cs

```diff
@@ -0,0 +1,364 @@
+
+            using (var getObjectIdCommand = new SqlCommand(getObjectIdQuery, connection))
+            {
+                using (SqlDataReader reader = await getObjectIdCommand.ExecuteReaderAsync(cancellationToken))
+                {
+
+                    // TODO: Check if the below if-block ever gets hit.
+                    if (!await reader.ReadAsync(cancellationToken))
+                    {
+                        throw new InvalidOperationException($"Failed to determine the OBJECT_ID of the user table {this._userTable.FullName}");
+                    }
+
+                    object userTableId = reader.GetValue(0);
+
+                    if (userTableId is DBNull)
+                    {
+                        throw new InvalidOperationException($"Failed to determine the OBJECT_ID of the user table {this._userTable.FullName}. " +
+                            "Possibly the table does not exist in the database.");
+                    }
+
+                    return (int)userTableId;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets the names and types of primary key columns of the user's table.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">
+        /// Thrown if no primary keys are found for the user table. This could be because the user table does not have
+        /// any primary key columns.
+        /// </exception>
+        private async Task<IReadOnlyList<(string name, string type)>> GetPrimaryKeyColumnsAsync(SqlConnection connection, CancellationToken cancellationToken)
+        {
+            string getPrimaryKeyColumnsQuery = $@"
+                SELECT c.name, t.name, c.max_length, c.precision, c.scale
+                FROM sys.indexes AS i
+                INNER JOIN sys.index_columns AS ic ON i.object_id = ic.object_id AND i.index_id = ic.index_id
+                INNER JOIN sys.columns AS c ON ic.object_id = c.object_id AND ic.column_id = c.column_id
+                INNER JOIN sys.types AS t ON c.user_type_id = t.user_type_id
+                WHERE i.is_primary_key = 1 AND i.object_id = OBJECT_ID(N{this._userTable.QuotedName}, 'U');
+            ";
+
+            using (var getPrimaryKeyColumnsCommand = new SqlCommand(getPrimaryKeyColumnsQuery, connection))
+            {
+                using (SqlDataReader reader = await getPrimaryKeyColumnsCommand.ExecuteReaderAsync(cancellationToken))
+                {
+
+                    string[] variableLengthTypes = new string[] { "varchar", "nvarchar", "nchar", "char", "binary", "varbinary" };
+                    string[] variablePrecisionTypes = new string[] { "numeric", "decimal" };
+
+                    var primaryKeyColumns = new List<(string name, string type)>();
+
+                    while (await reader.ReadAsync(cancellationToken))
+                    {
+                        string type = reader.GetString(1);
+
+                        if (variableLengthTypes.Contains(type))
```

Comment:
Can we guarantee these will always be lower case? Would suggest lower casing the value regardless to be safe.

Reply:
The rows in `sys.tables` should always have values under `Name` column in lower case at least for the system types. I will be updating the earlier check for conflicting column names to be case insensitive in next commit.

Reply:
Update on previous comment: The check for conflicting column name is case insensitive now.